### PR TITLE
Add OCI Vision MCP server

### PR DESCRIPTION
**Pull request by antimverma**
# Description

Adds the OCI Vision MCP server under `src/oci-vision-mcp-server`.

This server provides stdio MCP tools for OCI Vision image analysis and related Object Storage image workflows, including image analysis, OCR, object detection, face detection, async image jobs, Object Storage upload/list/fetch, stored result lookup, and runtime configuration status.

The implementation follows the repository MCP server structure and public packaging expectations:

- Adds `pyproject.toml`, `uv.lock`, `README.md`, `CHANGELOG.md`, `LICENSE.txt`, `Containerfile`, and `oracle/` package source.
- Exposes the package entry point as `oracle.oci-vision-mcp-server`.
- Supports stdio transport only.
- Uses OCI CLI-compatible session-token authentication only.
- Documents that HTTP, streamable HTTP, OAuth, IDCS bearer-token validation, `/mcp`, and `/.well-known/*` endpoints are not exposed.
- Adds OCI SDK `additional_user_agent` telemetry derived from package metadata.
- Includes unit/integration tests with coverage threshold set to 90%.

Fixes: N/A

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Validated locally from `src/oci-vision-mcp-server`:

- [x] `uv sync --locked --all-extras --dev`
- [x] `uv run pytest --cov=. --cov-branch --cov-report=term-missing`
  - `236 passed, 2 skipped`
  - Coverage: `91.31%`
- [x] `uv build`
  - Successfully built sdist and wheel.
- [x] Wheel content inspection
  - Confirmed wheel metadata and console entry point are present.
  - Confirmed test files are excluded from the wheel.
- [x] MCP client smoke testing with Codex using the locally built wheel
  - Registered the server as a local stdio MCP server.
  - Verified `detect_faces` completed successfully on a local image.
  - Verified `detect_text` completed successfully on a local screenshot and returned OCR lines.
- [x] Best-practices validation
  - UPL headers present on Python source and `Containerfile`.
  - `pyproject.toml` metadata and entry point validated.
  - stdio-only behavior documented and tested.
  - OCI SDK additional user-agent derivation tested.

**Test Configuration**:
- OS: macOS
- Python: 3.13.2
- Package manager: uv
- MCP transport: stdio
- OCI auth mode: OCI CLI-compatible session-token profile

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules (N/A)

### CODE REVIEW

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+    _allow_missing_required: bool = False,
+    _diagnostic_errors: list[dict[str, Any]] | None = None,
+) -> ResolvedMcpConfig:
+    # Kept for compatibility with older launcher calls; config now comes only
+    # from process environment variables and code defaults.
+    del persist_generated_profile
+    sources: dict[str, str] = {}
+    locked: dict[str, bool] = {}
+
+    profile = _required_env(
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py]: [P1] `src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py:379` — Optional configuration is enforced globally as required

- Evidence: `get_resolved_config()` unconditionally calls `_required_env()` for profile, region, and default compartment. Reproductions with an explicit tool-level compartment—or an Object Storage request with explicit namespace and bucket—still failed with `OCI_VISION_DEFAULT_COMPARTMENT_ID is required`. The README documents profile and region as optional and the compartment as unnecessary when supplied to a tool.
- Impact: The documented default-profile setup does not start, and Object Storage or local stored-result operations are blocked by an unrelated Vision compartment requirement.
- Suggested direction: Default the profile to `DEFAULT`, derive region from the OCI profile when not overridden, and validate the compartment only at Vision operations that actually need one.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/consts.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+DEFAULT_RUNTIME_DIR = Path.home() / ".oci-vision-mcp"
+DEFAULT_SESSION_AUTH_COMMAND = "oci"
+SESSION_AUTH_COMMAND_ENV = "OCI_SESSION_AUTH_COMMAND"
+DEFAULT_MAX_IMAGE_BYTES = 10 * 1024 * 1024
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/consts.py]: [P1] `src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/consts.py:14` — Accepted images exceed OCI Vision’s service contract

- Evidence: Local validation accepts GIF, BMP, TIFF, and WebP and allows images up to 10 MiB. OCI Vision image-analysis models accept only JPEG/PNG and up to 5 MB per image, according to the [OCI Vision limits](https://docs.oracle.com/en-us/iaas/Content/vision/using/limits.htm). Tests currently assert that the unsupported formats are accepted.
- Impact: Inputs accepted by the MCP server will deterministically fail at OCI, producing avoidable service errors and inconsistent behavior between local validation and execution.
- Suggested direction: Enforce JPEG/PNG and the 5 MB limit for Vision analysis. If Object Storage upload intentionally supports additional formats or sizes, give that tool a separate policy.

File: src/oci-vision-mcp-server/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+version = "0.1.0"
+description = "Stdio MCP server for OCI Vision and Object Storage image workflows."
+readme = "README.md"
+requires-python = ">=3.13"
+license = "UPL-1.0"
+license-files = ["LICENSE.txt"]
+authors = [
+    {name = "Oracle MCP", email = "237432095+oracle-mcp@users.noreply.github.com"},
+]
+dependencies = [
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/pyproject.toml:16]: Adding this here since its a project dependency, but its related to auth. Recently we added the `oracle-mcp-common` library with shared auth utils to keep things consistent in the MCP servers. Please review implementations in [oci-cloud-mcp-server ](https://github.com/oracle/mcp/blob/main/src/oci-cloud-mcp-server/oracle/oci_cloud_mcp_server/server.py#L480)and update this servers auth methods to utilize those shared utilities. If there are any gaps in `oracle-mcp-common` we are happy to discuss further enhancements with a separate issue.

[P1] `src/oci-vision-mcp-server/pyproject.toml:16` — OCI authentication bypasses the required shared implementation

- Evidence: The server does not depend on `oracle-mcp-common>=0.1.0,<0.2.0`. `authentication/session_signer.py:38` instead reads OCI profiles and constructs a `SecurityTokenSigner` directly.
- Impact: The server duplicates credential resolution, supports a narrower authentication path, and misses shared fixes and precedence behavior required for OCI SDK-backed servers in this repository.
- Suggested direction: Use `oracle_mcp_common.build_auth_context()` with `AuthOptions` if session-token-only behavior is intentional, merge the derived user agent into its config, and pass its signer to each OCI client.

File: src/oci-vision-mcp-server/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+]
+
+[tool.pytest.ini_options]
+testpaths = ["oracle/oci_vision_mcp_server/tests"]
+pythonpath = ["."]
+markers = [
+    "live: opt-in tests that call real OCI services",
+]
+
+[tool.coverage.run]
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/pyproject.toml:60]: [P2] `src/oci-vision-mcp-server/pyproject.toml:53` — Coverage excludes safety-critical I/O modules

- Evidence: Both `io/image_validation.py` and `io/object_storage.py` are omitted from coverage. The configured report is 91.31%; an independent run including them remained above the gate at 90.43%, but `io/object_storage.py` was only 68.60% covered.
- Impact: Regressions in streaming size enforcement, overwrite protection, cleanup, and response-body handling can land without affecting the coverage gate.
- Suggested direction: Remove these exclusions and add focused tests for streaming limits, overwrite refusal, temporary-file cleanup, and Object Storage response variants.

There should be coverage on those directories, or a good reason for excluding them should be documented in the README. `io/object_storage.py` should probably have more code coverage as well.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
This is a substantial and well-tested new server, but I found three blockers before merge:

1. `get_resolved_config()` requires profile, region, and default compartment globally, contradicting the documented defaults and blocking tools that do not need a Vision compartment.
2. Image validation accepts formats and sizes OCI Vision does not support.
3. OCI authentication bypasses the repository-required `oracle-mcp-common` authentication context.

Please also remove the coverage exclusions for the image-validation and Object Storage modules and add coverage for their safety-sensitive failure paths.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py (hunk 1 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py]: Addressed.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/consts.py (hunk 2 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/consts.py]: Addressed.

File: src/oci-vision-mcp-server/pyproject.toml (hunk 3 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/pyproject.toml:16]: Addressed.

File: src/oci-vision-mcp-server/pyproject.toml (hunk 4 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/pyproject.toml:60]: Addressed.

**Comment by antimverma** [comment]
Hi @dustin-sale Thanks for the review, I have addressed all your comments. Please check.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/session_signer.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    *,
+    profile: str | None = None,
+    region: str | None = None,
+) -> SessionAuthenticationError | None:
+    if not _is_auth_failure(exc):
+        return None
+
+    resolved_config = None if profile else get_resolved_config()
+    selected_profile = profile or resolved_config.profile
+    selected_region = region or resolved_config.region
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/session_signer.py]: [P1] Preserve profile-region fallback during 401 handling

- Evidence: When `profile` is supplied, line 86 sets `resolved_config` to `None`, but this line dereferences `resolved_config.region` whenever `region` is omitted. The README explicitly permits omitting `OCI_REGION` and using the selected profile's region. Calling this handler for a 401 with `profile="DEFAULT", region=None` reproduces `AttributeError: 'NoneType' object has no attribute 'region'`.
- Impact: An expired or rejected session under the documented default setup escapes the tool's structured, retryable `OCI_SESSION_AUTH_REQUIRED` response.
- Requested change: Handle the missing explicit region without dereferencing `None`, and add a regression test that sends a profile with `region=None` through a service-error envelope.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        decoded = base64.urlsafe_b64decode(f"{payload}{padding}")
+        data = json.loads(decoded)
+    except (binascii.Error, ValueError, json.JSONDecodeError):
+        return None
+    exp = data.get("exp")
+    return exp if isinstance(exp, int) else None
+
+
+def _run_session_refresh(profile: str, *, executable: str) -> None:
+    command = [executable, "session", "refresh", "--profile", profile]
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: [P1] Propagate custom OCI config paths to session repair

- Evidence: `_load_session_config()` now reads `resolve_config_file()`, but both subprocess commands omit that path. OCI CLI consumes `OCI_CLI_CONFIG_FILE`, not the server's documented `OCI_CONFIG_FILE`. Capturing both helpers with `OCI_CONFIG_FILE=/private/tmp/custom-oci-config` produced commands without `--config-file` or `--config-location`.
- Impact: An expired session loaded from a custom config is repaired against the default config instead, potentially failing or updating the wrong profile.
- Requested change: Pass `resolve_config_file()` explicitly to refresh and authentication using the appropriate OCI CLI options, and assert both generated commands in tests.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/schemas.py [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+                raise ValueError("metadata keys and values must not contain line breaks.")
+        return value
+
+    @model_validator(mode="after")
+    def validate_upload_source(self) -> "ObjectStorageUploadInput":
+        if (self.image is None) == (self.images is None):
+            raise ValueError("Provide exactly one of image or images.")
+        selected_images = [self.image] if self.image is not None else list(self.images or [])
+        for image in selected_images:
+            if image.source_type != ImageSourceType.FILE_PATH:
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/schemas.py]: [P2] Publish an accurate upload source contract

- Evidence: The upload model exposes the shared `ImageInput` schema, which advertises base64, file-path, OCI-object, and URL sources, while this validator accepts only file paths. The README also describes the tool as supporting local/base64 upload.
- Impact: MCP clients can construct schema-valid base64 requests that always fail at runtime.
- Requested change: Use a file-path-only schema for this tool, or implement base64 upload, and align the README with the resulting contract.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+        effect=(
+            "When true, startup and tool calls may run "
+            "`<session_auth_command> session refresh --profile <profile>`."
+        ),
+    ),
+    EnvVarInfo(
+        name=ENV_AUTO_AUTH,
+        purpose="Allow startup to run browser-based session authentication.",
+        required=False,
+        default="true",
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py]: [P3] Correct the auto-auth catalog default

- Evidence: Runtime configuration and the README default auto-auth to false, but this catalog entry reports `"true"`; `get_config_status` therefore returns contradictory metadata.
- Impact: Clients inspecting configuration status can incorrectly conclude that browser authentication is enabled by default.
- Requested change: Derive this value from `DEFAULT_AUTO_AUTH`, as the other catalog defaults do.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
## Requested changes
- `authentication/session_signer.py:88` — [P1] Preserve profile-region fallback during 401 handling.
- `authentication/auth.py:124` — [P1] Propagate custom OCI config paths to session-repair commands.

## Additional review notes
- `config/schemas.py:331` — [P2] Publish an accurate upload source schema.
- `config/settings.py:209` — [P3] Correct the auto-auth catalog default.

## Validation
- `make lint` passed.
- `make test project=oci-vision-mcp-server` passed: 251 passed, 2 skipped, 90.86% coverage.
- The two authentication edge failures were reproduced independently.

See the inline comments in this review for evidence, impact, and suggested remediation.

**Comment by antimverma** [comment]
HI @dustin-sale, addressed all the open comments. Please check.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    command = [
+        executable,
+        "session",
+        "authenticate",
+        "--profile-name",
+        profile,
+    ]
+    command.extend(["--region", region])
+    if config_file:
+        command.extend(["--config-file", config_file])
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: [P1] Use the authentication config destination option

- Evidence: OCI CLI 3.90.0 parses the added `--config-file` as its global read-config option. A safe invocation using this command shape left the `session authenticate` callback's `config_location` null, while replacing it with `--config-location` supplied `/private/tmp/custom-oci-config`. The CLI persists the generated session using `config_location`.
- Impact: when `OCI_CONFIG_FILE` points to a custom file, automatic authentication writes the new profile elsewhere. The server then reloads the unchanged custom file and remains unauthenticated. The recovery commands built here and in `session_signer.py` direct users to the wrong destination too.
- Requested change: use `--config-location` for `_run_session_authenticate`, `_session_authenticate_command`, and `session_auth_command`. Keep `--config-file` for refresh, and update the tests to assert the distinct flags.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py [hunk 10]
```diff
@@ -0,0 +1,905 @@
...
+        required=False,
+        default=str(DEFAULT_MAX_IMAGE_BYTES),
+        used_in="config/settings.py, io/image_loader.py",
+        effect="Images larger than this are rejected before OCI calls.",
+    ),
+    EnvVarInfo(
+        name=ENV_REFRESH_SESSION,
+        purpose="Allow startup to refresh expired OCI session tokens.",
+        required=False,
+        default=str(DEFAULT_AUTO_AUTH).lower(),
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/config/settings.py]: [P2] Align both catalog defaults with runtime defaults

- Evidence: `DEFAULT_REFRESH_SESSION` is `True` and `DEFAULT_AUTO_AUTH` is `False`, but this refresh entry now derives from `DEFAULT_AUTO_AUTH` while the following auto-auth entry still declares `"true"`.
- Impact: `get_config_status` reports refresh as disabled and browser-based auto-auth as enabled, the reverse of runtime behavior.
- Requested change: derive the refresh entry from `DEFAULT_REFRESH_SESSION`, derive the auto-auth entry from `DEFAULT_AUTO_AUTH`, and add a regression assertion for both catalog values.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
## Requested changes
- `authentication/auth.py:153` — [P1] Use the authentication config destination option.

## Additional review notes
- `config/settings.py:198` — [P2] Align both catalog defaults with runtime defaults.

## Validation
- `make lint` and `make test project=oci-vision-mcp-server` passed; 252 tests passed, 2 skipped, with 90.77% coverage.
- OCI CLI 3.90.0 parsing confirmed that `--config-file` does not populate the authentication destination.

See the inline comments in this review for evidence, impact, and suggested remediation.

**Comment by antimverma** [comment]
@dustin-sale Addressed the open comments.

File: src/oci-vision-mcp-server/pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+description = "Stdio MCP server for OCI Vision and Object Storage image workflows."
+readme = "README.md"
+requires-python = ">=3.13"
+license = "UPL-1.0"
+license-files = ["LICENSE.txt"]
+authors = [
+    {name = "Oracle MCP", email = "237432095+oracle-mcp@users.noreply.github.com"},
+]
+dependencies = [
+    "fastmcp==3.4.2",
```
Comment (dustin-sale) [inline src/oci-vision-mcp-server/pyproject.toml]: Can you update these dependencies to the following:

- fastmcp==3.4.5
- Remove mcp since that is a sub-dependency unless the code is importing from the mcp package. If its needed, update it to mcp>=1.29.0,<2.0.0
- oci==2.182.1 
- pydantic>=2.13.4

Just trying to get ahead of a few known security vulnerabilities that will come up once this gets merged.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
Previous requested changes have all been addressed.

I am ready to approve once the dependencies are updated, see my comment.

File: src/oci-vision-mcp-server/pyproject.toml (hunk 11 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/pyproject.toml]: Addressed

**Review by dustin-sale — APPROVED** [review]
Approved. @gebhardtr will need to review as well.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        region = config.region or resolved_config.region
+        if not region:
+            raise RuntimeError(
+                "OCI region is required before automatic session authentication can run. "
+                "Set OCI_REGION in your MCP client environment and restart the MCP server."
+            )
+        _run_session_authenticate(
+            config.profile,
+            region,
+            executable=resolved_config.session_auth_command,
```
Comment (gebhardtr) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: should the mcp server handle re-authentication?  this feels like a separation of concerns issue

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    profile: str, *, executable: str, config_file: str | None = None
+) -> None:
+    command = [executable, "session", "refresh", "--profile", profile]
+    if config_file:
+        command.extend(["--config-file", config_file])
+    # Refresh can fail for expired sessions; authenticate may still repair it.
+    subprocess.run(command, check=False, stdin=sys.stdin, stdout=sys.stderr, stderr=sys.stderr)
+
+
+def _run_session_authenticate(
```
Comment (gebhardtr) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: this introduces another dependency, another surface to secure and limits how it can be deployed. we shouldn't invoke subprocesses

**Review by gebhardtr — CHANGES_REQUESTED** [review]
Auth/subprocess invocation comments.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py (hunk 12 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: @gebhardtr Addressed this.

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py (hunk 13 above)
Comment (antimverma) [reply src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/authentication/auth.py]: @gebhardtr Yes agreed. This is addressed.

**Comment by antimverma** [comment]
> Auth/subprocess invocation comments.

@gebhardtr This is addressed. Please review/approve.

File: src/oci-vision-mcp-server/Containerfile [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+RUN microdnf install epel-release && \
+  microdnf install python3.13 python3.13-pip && \
+  rm -rf /var/cache/dnf/* && \
+  useradd -m -d /app oracle
+
+WORKDIR /app
+COPY --chown=oracle:oracle . /app
+
+RUN pip3.13 install --no-cache-dir uv && \
+  uv --no-cache sync --locked --all-extras
```
Comment (gebhardtr) [inline src/oci-vision-mcp-server/Containerfile]: doesn't this break when attempting to sync the `common` package?

File: src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/tests/test_runtime_logging.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    tee.flush()
+    assert primary.getvalue() == "hello"
+    assert secondary.getvalue() == "hello"
+    assert tee.isatty() is False
+    assert tee.encoding is None
+
+
+def test_configure_stderr_log_is_idempotent(monkeypatch, tmp_path) -> None:
+    original_stderr = sys.stderr
+    first_log_path = configure_stderr_log(str(tmp_path / "logs"))
```
Comment (gebhardtr) [inline src/oci-vision-mcp-server/oracle/oci_vision_mcp_server/tests/test_runtime_logging.py]: are we using this anywhere? apologies if I missed it, but I don't see calls to this function outside of tests.

**Comment by antimverma** [comment]
@gebhardtr All comments addressed. Can you please review.
